### PR TITLE
osd: fix C++20 atomic deprecation warnings

### PR DESCRIPTION
**Pull request by chardan**
gcc14 issues warnings about C++11 atomic features that were deprecated in C++20. This
patch migrates the types involved to atomic<shared_ptr<>> to resolve this.

note:
Does NOT patch for WITH_SEASTAR=ON.

refer-to:
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2869r0.pdf

Fixes: https://tracker.ceph.com/issues/69026

Signed-off-by: Jesse F. Williamson <nerd.cpp@gmail.com>

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/osd/osd_types.h [hunk 1]
```diff
@@ -3497,7 +3497,7 @@ class PastIntervals {
       old_up_primary, new_up_primary,
       old_up, new_up,
       same_interval_since, last_epoch_clean,
-      osdmap.get(), lastmap.get(),
+      osdmap.load(), lastmap.load(),
```
Comment (Matan-B) [inline src/osd/osd_types.h:3500]: See make check:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/osd_types.h:3487:15: note: in instantiation of template class 'std::atomic<std::shared_ptr<const OSDMap>>' requested here
    OSDMapRef osdmap,      ///< [in] current map
              ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/osd_types.h:3500:30: error: no member named 'load' in 'std::atomic<std::shared_ptr<const OSDMap>>'
      osdmap.load(), lastmap.load(),
                     ~~~~~~~ ^
2 errors generated.
```

**Review by Matan-B — COMMENTED** [review]
Regarding `WITH_SEASTAR`, it depends on the case. Given that the change here is general and not Classical OSD specific, I think that we should also adjust Crimson's relevant users (if any).

File: src/osd/osd_types.h (hunk 1 above)
Comment (chardan) [reply src/osd/osd_types.h:3500]: Thanks  for taking a look-- I'm not sure how I missed this in my local build!

**Comment by cbodley** [comment]
unfortunately, we're currently stuck supporting gcc 11 which doesn't have atomic shared ptr yet:
```
include/c++/11/atomic:211:7: error: static_assert failed due to requirement '__is_trivially_copyable(std::shared_ptr<const OSDMap>)' "std::atomic requires a trivially copyable type"
      static_assert(__is_trivially_copyable(_Tp),
      ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
this came up recently in https://github.com/ceph/ceph/pull/60225, where we needed ifdefs for gcc version

**Comment by chardan** [comment]
Ah, shucks! That complicates this. We /could/ work around it with what amount to a wrapper for atomic_store() that does a capability check (#if __cpp_lib_atomic_shared_ptr), but unless we're doing it project wide (and that seems like somewhat pointless effort as the "right" solution is to move to atomic<shared_ptr<>>) I feel like it adds quite a bit of complexity to an already trickier-than-it-looks[1] bit of code. Perhaps it's best to let this one be for now.

1. Noticing that OSDMap.h (1838) has its /own/ Seastar ifdef of OSDMapRef for the global namespace, which is different than the PastIntervals::OSDMapRef appearing in osd_types.h, and looking at the way the tests use it, it becomes curious that the overload interval_rep::check_new_interval() appears to only be there FOR wrangling the atomic types... to pass the values into a function that doesn't do atomic checks. I get a bit unclear on what was actually intended .

**Comment by chardan** [comment]
Closing this for now-- it will make more sense to address once we can drop support for GCC11.

See-also: https://lists.ceph.io/hyperkitty/list/dev@ceph.io/thread/LNMXUFBMFXHIXX334DYOMW5K4PYN4YIM/

**Pull request closed without merging**